File: 35_Even_Wins/python/evenwins.py
# evenwins.py

# global variables
total = 0
m = 0
m1 = 0
go_first = ''
human_move = 0

intro_text = """
EVEN WINS
CREATIVE COMPUTING  MORRISTOWN, NEW JERSEY

     THIS IS A TWO PERSON GAME CALLED 'EVEN WINS.'
TO PLAY THE GAME, THE PLAYERS NEED 27 MARBLES OR
OTHER OBJECTS ON A TABLE.

     THE ONLY RULES ARE THAT (1) YOU MUST ALTERNATE TURNS,
(2) YOU MUST TAKE BETWEEN 1 AND 4 MARBLES EACH TURN,
AND (3) YOU CANNOT SKIP A TURN.

     THE 2 PLAYERS ALTERNATE TURNS, WITH EACH PLAYER
REMOVING FROM 1 TO 4 MARBLES ON EACH MOVE.  THE GAME
ENDS WHEN THERE ARE NO MARBLES LEFT, AND THE WINNER
IS THE ONE WITH AN EVEN NUMBER OF MARBLES.

     TYPE A '1' IF YOU WANT TO GO FIRST, AND TYPE
A '0' IF YOU WANT ME TO GO FIRST.
"""

# 1160 PRINT
# 1170 PRINT "THE NUMBER OF MARBLES YOU TAKE MUST BE A POSITIVE"
# 1180 PRINT "INTEGER BETWEEN 1 AND 4."
# 1190 PRINT
# 1200 PRINT "     WHAT IS YOUR NEXT MOVE?"
# 1210 PRINT
bad_move_text = """
THE NUMBER OF MARBLES YOU TAKE MUST BE A POSITIVE
INTEGER BETWEEN 1 AND 4.

     WHAT IS YOUR NEXT MOVE?
"""

def bad_move(): print(bad_move_text)

def print_total(): print(f'\nTOTAL={total}')


def game_over():
    # 880 PRINT "THAT IS ALL OF THE MARBLES."
    # 890 PRINT
    # 900 PRINT " MY TOTAL IS";M1;", YOUR TOTAL IS";Y1
    # 910 PRINT
    # 920 IF INT(M1/2)=M1/2 THEN 950
    # 930 PRINT "     YOU WON.  DO YOU WANT TO PLAY"
    print('THAT IS ALL OF THE MARBLES.')
    print()
    print(f' MY TOTAL IS {m1}, YOUR TOTAL IS {y1}')
    print()
    if int(m1/2) == m1/2:
        # 950 PRINT "     I WON.  DO YOU WANT TO PLAY"
        print('     I WON.  DO YOU WANT TO PLAY')
    else:
        print('     YOU WON.  DO YOU WANT TO PLAY')
    # 960 PRINT "AGAIN?  TYPE 1 FOR YES AND 0 FOR NO."
    # 970 INPUT A1
    print('AGAIN?  TYPE 1 FOR YES AND 0 FOR NO.')
    play_again = input('? ')
    if play_again == '0':
        print('OK.  SEE YOU LATER.')
        return

def main():
    print(intro_text)
    go_first = input('? ')
    if go_first == '0':
        # computer goes first
        total = 27  # start with 27 marbles
        m = 2
        # 270 PRINT:PRINT "TOTAL=";T:PRINT
        # 280 M1=M1+M
        # 290 T=T-M
        # 300 PRINT "I PICK UP";M;"MARBLES."
        print_total()
        m1 = m1 + m
        total = total - m
        if total == 0:
            game_over()
        else: #  there are 1 or more marbles left   
            # 320 PRINT:PRINT "TOTAL=";T
            # 330 PRINT
            # 340 PRINT "     AND WHAT IS YOUR NEXT MOVE, MY TOTAL IS";M1
            # 350 INPUT Y
            # 360 PRINT
            print_total()
            print()
            print(f'     AND WHAT IS YOUR NEXT MOVE, MY TOTAL IS {m1}')
            human_move = input('? ')
            # 370 IF Y<1 THEN 1160
            # 380 IF Y>4 THEN 1160
            if human_move < 1: bad_move()
            if human_move > 4: bad_move()

            # 390 IF Y<=T THEN 430


    else:
        # human goes first


if __name__ == '__main__':
    main()

File: 35_Even_Wins/cpp/evenwins.cpp
// evenwins.cpp

#include <iostream>
#include <string>
#include <ctime>
#include <cstdlib>

using namespace std;

// global variables
int marbles_in_middle = -1;
int human_marbles     = -1;
int computer_marbles  = -1;
string whose_turn     = "";

void welcome_screen() {
    cout << "Welcome to Even Wins!\n";
    cout << "Based on evenwins.bas from Creative Computing\n";
    cout << "\n";
    cout << "Even Wins is a two-person game. You start with\n";
    cout << "27 marbles in the middle of the table.\n";
    cout << "\n";
    cout << "Players alternate taking marbles from the middle.\n";
    cout << "A player can take 1 to 4 marbles on their turn, and\n";
    cout << "turns cannot be skipped. The game ends when there are\n";
    cout << "no marbles left, and the winner is the one with an even\n";
    cout << "number of marbles.";
    cout << "\n";
}

string marbles_str(int n) {
    if (n == 1) return "1 marble";
    return to_string(n) + " marbles";
}

void choose_first_player() {
   for (;;) {
        cout << "\nDo you want to play first? (y/n) --> ";
        string ans;
        cin >> ans;
        if (ans == "y") {
            whose_turn = "human";
            return;
        } else if (ans == "n") {
            whose_turn = "computer";
            return;
        } else {
            cout << "\nPlease enter 'y' if you want to play first,\n";
            cout << "or 'n' if you want to play second.\n";
        }
    }
} // choose_first_player

void next_player() {
    if (whose_turn == "human"){
        whose_turn = "computer";
    } else if (whose_turn == "computer") {
        whose_turn = "human";
    } 
}

void print_board() {
    cout << "\n";
    cout << " marbles in the middle: " << marbles_in_middle 
         << " " << string(marbles_in_middle, '*') << "\n";
    cout << "    # marbles you have: " << human_marbles << "\n";
    cout << "# marbles computer has: " << computer_marbles << "\n";
    cout <<  "\n";
}

// returns true if s is an int, and false otherwise
// https://www.cplusplus.com/reference/string/stoi/
bool is_int(const string& s) {
    try {
        stoi(s); // throws an invalid_argument exception if conversion fails
        return true;
    } catch (...) {
        return false;
    }
}

void human_turn() {
    // get number in range 1 to min(4, marbles_in_middle)
    int max_choice = min(4, marbles_in_middle);
    cout << "It's your turn!\n";
    for(;;) {
        cout << "Marbles to take? (1 - " << max_choice << ") --> ";
        string s;
        cin >> s;
        if (!is_int(s)) {
            cout << "\n  Please enter a whole number from 1 to " << max_choice
                 << "\n";
            continue;
        }

        // convert s to an int so it can be compared
        int n = stoi(s);
        if (n < 1) {
            cout << "\n  You must take at least 1 marble\n";
            continue;
        }
        if (n > max_choice) {
            cout << "\n  You can take at most " << marbles_str(max_choice)
                 << "\n";
            continue;
        }

        cout << "\nOkay, taking " << marbles_str(n) << " ...\n";
        marbles_in_middle -= n;
        human_marbles += n;

        return;
    } // for
} // human_turn

void computer_turn() {
    cout << "It's the computer's turn ...\n";
    int max_choice = min(4, marbles_in_middle);

    // choose at random
    int n = 1 + rand() % max_choice;
    cout << "Computer takes " << marbles_str(n) << " ...\n";
    marbles_in_middle -= n;
    computer_marbles += n;
}

void game_over() {
    cout << "\n";
    cout << "!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!\n"
         << "!! All the marbles are taken: Game Over !!\n"
         << "!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!\n\n";
    print_board();
    if (human_marbles % 2 == 0) {
        cout << "You are the winner! Congratulations!\n";
    } else {
        cout << "The computer wins: all hail mighty silicon!\n";
    }
}

void play_game() {
    // initialize the game state
    marbles_in_middle = 27;
    human_marbles = 0;
    computer_marbles = 0;
    print_board();

    for (;;) {
        if (marbles_in_middle == 0) {
            game_over();
            return;
        } else if (whose_turn == "human") {
            human_turn();
            print_board();
            next_player();
        } else if (whose_turn == "computer") {
            computer_turn();
            print_board();
            next_player();
        }
    } // for
} // play_game

int main() {
    srand(time(NULL));

    welcome_screen();

    for(;;) {
        choose_first_player();
        play_game();

        // ask if the user if they want to play again
        cout << "\nWould you like to play again? (y/n) --> ";
        string again;
        cin >> again;
        if (again == "y") {
            cout << "\nOk, let's play again ...\n";
        } else {
            cout << "\nOk, thanks for playing ... goodbye!\n";
            return 0;
        }
    } // for
} // main
